File: backend/test-solutions/word-break-ii.cpp
#include <vector>
#include <string>
#include <unordered_set>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> wordSet(wordDict.begin(), wordDict.end());
        unordered_map<int, vector<string>> memo;
        
        return dfs(s, 0, wordSet, memo);
    }
    
private:
    vector<string> dfs(const string& s, int start, const unordered_set<string>& wordSet, 
                       unordered_map<int, vector<string>>& memo) {
        if (memo.find(start) != memo.end()) {
            return memo[start];
        }
        
        vector<string> result;
        
        if (start == s.length()) {
            result.push_back("");
            return result;
        }
        
        for (int end = start + 1; end <= s.length(); end++) {
            string word = s.substr(start, end - start);
            if (wordSet.find(word) != wordSet.end()) {
                vector<string> rest = dfs(s, end, wordSet, memo);
                for (const string& sentence : rest) {
                    if (sentence.empty()) {
                        result.push_back(word);
                    } else {
                        result.push_back(word + " " + sentence);
                    }
                }
            }
        }
        
        memo[start] = result;
        return result;
    }
};

File: backend/test-solutions/word-break-ii.js
class Solution {
  wordBreak(s, wordDict) {
    const wordSet = new Set(wordDict);
    const memo = new Map();

    function dfs(start) {
      if (memo.has(start)) {
        return memo.get(start);
      }

      if (start === s.length) {
        return [""];
      }

      const result = [];
      for (let end = start + 1; end <= s.length; end++) {
        const word = s.substring(start, end);
        if (wordSet.has(word)) {
          const rest = dfs(end);
          for (const sentence of rest) {
            if (sentence) {
              result.push(word + " " + sentence);
            } else {
              result.push(word);
            }
          }
        }
      }

      memo.set(start, result);
      return result;
    }

    return dfs(0);
  }
}
